### ISSUE TITLE
description of architecture of compiler, runtime, etc?

### ISSUE THREAD

**Issue by dominictarr** [alantech/alan#241]
A friend linked me to alan after I was ranting about how everyone does async wrong (explicit await or callbacks when it can be inferred) I was also very excited to see that alan has node.js inspired modules!

I'm currently also dabbling on my own language, and would like to dive in and see how alan does stuff, but it would be very helpful if there was at least a rough description of how the compiler works. I have browsed the code, but there were not many comments, but I saw that there are at least two formats, amm and agc, as well as the text source code. I also saw that the issue about documentation: https://github.com/alantech/alan/issues/223

but in the mean time it would just be very interesting to have a rough description of how it's structured. What features does the runtime support? what does raw amm look like? what are the most interesting transformations? (for example to infer the async dependency graph)

**Comment by dfellis** [comment]
Hi @dominictarr sorry about that documentation state. We focused on getting docs for people who want to use it prioritized since it explains what we're trying to do and they're probably the majority of visitors. We will definitely clean that up over time, especially after we remove ANTLR from our compiler and shrink its size by multiple megabytes of generated source code.

The compile has multiple intermediate layers with a branching path for the two runtime targets: Javascript and our native runtime written in Rust.

```
       LN (aLaN)
        |
       AMM (Alan Minus Minus, simplified syntax)
      /   \
   JS      AGA (Alan Graphcode Assembler, dependency graph code in a text format)
            |
           AGC (Alan Graphcode, binary format for the runtime)
```

**LN** is the high level language meant to look a lot like a cleaned-up Typescript/JS. Automatic memory management, automatic IO concurrency, automatic event-level and array-level CPU parallelism (two tracks in the runtime right now), and a strict type system but with interfaces for defining generic functions that we believe we can write [100% automatic type inference with without the compiler requiring a fallback to an explicit type](https://github.com/alantech/alan/blob/main/rfcs/006%20-%20Automatic%20Argument%20Interfaces%20RFC.md).

**AMM** is a heavily stripped-down version of the language but technically syntactically compatible. It only has constants, events, event handlers, functions, closures, function calls, and assignments. Everything else, including conditionals, is stripped away. So individual statements in *ln* files are broken into multiple of these "microstatements" and that's where the majority of the compiler logic lives, doing the AST manipulation into this. It also requires constants to be hoisted to the outermost scope and does constant deduplication to minimize output size later on.

Since functions in Alan are really just acyclic directed graphs (no explicit loops or recursion) we also *massively* inline functions wherever possible. Inlining stops when the compiler runs into an "opcode", a unitary function that for the needs of the runtime is limited to 8 letters and 3 total arguments (2-in 1-out or 3-in 0-out). These are strictly typed and so all of the type inference in the language builds off of them.

**JS:** We noticed AMM looked surprisingly close to Javascript/Typescript, so we wrote a transpiler that mostly just strips out the type information and replaces all of opcode calls with method calls on [a pretty simple runtime shim](https://github.com/alantech/alan/blob/main/js-runtime/index.js) (that we licensed Apache 2.0 so there's no problems redistributing it with the compiler output).

The transpiled Javascript doesn't have any of the benefits of the language in terms of cpu parallelism or automatic concurrency, but we think we could eventually add those things back with some extra intelligence in the transpilation and judicious use of WebWorker threads, but we're also not sure if we're going to continue down that path or target something like AssemblyScript, instead.

**AGA** is the layer that converts the AMM into something that looks very similar to assembly, but with annotation on events, closures, and the dependency graph. This is the layer that does the dependency graph analysis, but it has some bugs at the moment. The runtime doesn't use the dependency graph for too much at the moment so we implemented the parts that we need, but there is [a ticket to resolve this](https://github.com/alantech/alan/issues/197) that we hope to get to soon.

I personally find this version more legible to read than the AMM layer, but that's mostly due to the terrible names the first layer of the compiler comes up with right now (and some indent issues with deeply-nested closures).

**AGC** is the binary version of AGA and we could probably have written it with a single `sed` regex if we were insane enough (I was tempted). There's not too much interesting stuff there and it's probably the most "done" layer of the compiler.

**Runtime:** What the runtime supports right now is: all of the opcodes except the HTTP server-related ones ([working on that right now!](https://github.com/alantech/alan/pull/239)). The array opcodes automatically parallelize the closure they're executing, but we've also exposed the linear variants because we don't have automatic de-parallelization via functional purity testing working yet. We may keep those exposed for people who *don't* want them to grab multiple CPUs while doing their work. (We expect early on there could be throughput clashing between the array parallelism and event parallelism until we get those two systems interlinked and have the runtime making those trade-off decisions.)

The event-level parallelism is pretty simple, each event handler is effectively the entrypoint for its "own" mini program in the runtime and the event value being the "argc/v" equivalent. Both of these compute parallelisms currently use [Rayon](https://github.com/rayon-rs/rayon), though we think we may eventually outgrow it since we have more information than Rayon does on the workloads involved (we should be able to predict the compute time of the closure with some error bar, wider for closures with conditionals, coupled with the size of the data and cost to scatter to threads and marshal back, while Rayon can only use heuristics).

The IO concurrency is done through the [Tokio](https://tokio.rs/) event loop (which also handles the triggering of the events) and the runtime uses the AGC dependency graph declarations to determine which IO operations are safe to group together and execute in parallel, and which needs to be done sequentially. This is accomplished by having three classes of opcodes: `cpu`, `unpred_cpu`, and `io`, with `cpu` being very analogous to instructions in a CPU, `unpred_cpu` being things that have argument-dependent running times, like `array.map` that are usually parallelizable in Rayon, and `io` literally being IO operations.

**Modules** You didn't ask directly, but there's are tweaks to the module system in Alan versus Node.js that open up some really interesting stuff. The opcodes that the compiler works with are only exposed to the standard library, so having access to them is gated on eventually importing something that is telling the runtime what to do with it, with the root scope being an implicit import in every module.

The import process, however, allows you to effectively "shadow" other modules with the same name, which lets you do mocks "for free" in the language. The shadowing module can import the module it is replacing, so you can also alter the behavior, and this is even allowed for the standard library. So you could, for instance, shadow `@std/app` and replace the `print` function with a hook to your own logger library.

You could also replace a standard library module with a _non-functioning_ version of that module, and this can be scoped to a single module, including third-party dependencies (by having a two-headed dependency search path, `modules` and `dependencies`, with `modules` checked first, then `dependencies`). This makes it possible to literally block off parts of the runtime from third-party dependencies, so we should be able to integrate that with the package management system to provide a granular, per-module permissioning system and eliminate a class of vulnerabilities due to trust concerns (not completely, of course, but defense-in-depth-style).

I'm probably going to edit and expand this into new documentation, but wanted to answer your question relatively quickly. :)

**Comment by dominictarr** [comment]
thank you for your detailed reply! it gave me so much to say that I didn't get it all out, and while trying to build alan, and having to restart windows many times to get linux properly working for rust I lost what I had written. Well, I'll just say that I'm very excited about the features you mention (especially the module system, that's what originally attracked me to node.js, which I started using at version 0.2.1! and not enough people doing other languages realize how much better the node module system is)

Also, thanks those where the hints that I needed.

btw, if you are looking to replace antlr and since you are already working in javascript you might be interested in https://github.com/dominictarr/stack-expression it would be straightforward to port your syntax (because it's also based on *|+ etc) ... I hate to turn up in an issue and recommend my own module, so I won't feel bad if you don't use it, but I do think it's a problem no one seems to realize how simple parsing can be...

**Comment by dfellis** [comment]
Thank you @dominictarr :) I'm assuming from the above that you're running Linux in a VM? To help out with the other issue, would you mind telling me which Linux distribution you're using?

I will certainly take a look at your tool, though I will be upfront that it'll be an uphill battle. [We already wrote a replacement tool here](https://github.com/alantech/alan/blob/main/compiler/src/lp.ts) that we're using for all of the other layers of the compiler except the first one, mostly because that transition is a lower priority versus other needs of the language, and also because [my first attempt at doing that showed just how massive that migration is going to be](https://github.com/alantech/alan/tree/drop-antlr).

**Comment by dfellis** [comment]
The explanation in here has been expanded into documentation on the compiler internals and more rigor on the intermediate representations: https://docs.alan-lang.org/compiler_internals.html

Going to close this issue out, now.
